<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [259]:
import pandas as pd
import psycopg2

In [260]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [261]:
connection = psycopg2.connect(
    dbname = 'project_sql',
    user = 'skillfactory',
    password = 'cCkxxLVrDE8EbvjueeMedPKt',
    host = '84.201.134.129',
    port = 5432
)

In [262]:
# Hiding message of user warnings  
import warnings
warnings.filterwarnings('ignore', category=UserWarning)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [263]:
# текст запроса
query_3_1 = f'''
        select count(*)
        from vacancies
'''

In [264]:
# результат запроса
total_vacancies = pd.read_sql_query(query_3_1, connection)
total_vacancies = total_vacancies.iloc[0,0]
print('Количество вакансий:', total_vacancies)

Количество вакансий: 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [265]:
# текст запроса
query_3_2 = f'''
        select count(*)
        from employers
'''

In [266]:
# результат запроса
total_employers = pd.read_sql_query(query_3_2, connection)
total_employers = total_employers.iloc[0,0]
print('Количество работодателей:', total_employers)

Количество работодателей: 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [267]:
# текст запроса
query_3_3 = f'''
        select count(*)
        from areas
'''

In [268]:
# результат запроса
total_areas = pd.read_sql_query(query_3_3, connection)
total_areas = total_areas.iloc[0,0]
print('Количество регионов:', total_areas)

Количество регионов: 1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [269]:
# текст запроса
query_3_4 = f'''
        select count(*)
        from industries
'''

In [270]:
# результат запроса
total_industries = pd.read_sql_query(query_3_4, connection)
total_industries = total_industries.iloc[0,0]
print('Количество сфер деятельности:', total_industries)

Количество сфер деятельности: 294


***

**ВЫВОД ПО ПРЕДВАРИТЕЛЬНОМУ АНАЛИЗУ ДАННЫХ:**

*Количество сфер деятельности:* В базе данных содержится информация о 294 различных сферах деятельности. Это может указывать на разнообразие компаний и организаций, представленных в базе данных.

*Количество регионов:* В базе данных учитываются 1362 различных региона. Это может указывать на географическое распределение вакансий и возможное разнообразие рынков труда в разных регионах.

*Количество работодателей:* Существует информация о 23501 работодателе. Это количество компаний и организаций, предлагающих вакансии. Большое количество работодателей может указывать на активность рынка труда и разнообразие предложений о работе.

*Количество вакансий:* Общее количество вакансий составляет 49197. Это количество рабочих позиций, доступных в базе данных. Большое количество вакансий может указывать на востребованность рынка труда.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [345]:
# текст запроса
query_4_1 = f'''
        select a.name,
               count(v.id) cnt
        from vacancies v
        join areas a on v.area_id = a.id
        group by a.name
        order by 2 desc
        limit 5
'''

In [346]:
# результат запроса
cnt_vacancies = pd.read_sql_query(query_4_1, connection)
display(cnt_vacancies)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [273]:
# текст запроса
query_4_2 = f'''
        select count(id) cnt
        from vacancies 
        where salary_from is not null or salary_to is not null
'''

In [274]:
# результат запроса
not_null_vac = pd.read_sql_query(query_4_2, connection)
not_null_vac = not_null_vac.iloc[0,0]
print()
print('Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой: ', not_null_vac)


Количество вакансий, у которых заполнено хотя бы одно из двух полей с зарплатой:  24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [275]:
# текст запроса
query_4_3 = f'''
        select round(avg(salary_from)) as low_avg,
            round(avg(salary_to)) as high_avg
        from vacancies 
        
'''

In [276]:
# результат запроса
avg_salary = pd.read_sql_query(query_4_3, connection)
display(avg_salary)

low_avg  high_avg
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [277]:
# текст запроса
query_4_4 = f'''
        select schedule,
                employment,
                count(id)
        from vacancies
        group by schedule, employment
        order by 3 desc
        
'''

In [278]:
# результат запроса
cnt_vac_type = pd.read_sql_query(query_4_4, connection)
display(cnt_vac_type)

schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [279]:
# текст запроса
query_4_5 = f'''
        select experience,
            count(id)
        from vacancies
        group by experience
        order by 2
        
'''

In [280]:
# результат запроса
cnt_vac_exp = pd.read_sql_query(query_4_5, connection)
display(cnt_vac_exp)

experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***


**Вывод:**

*Распределение вакансий по городам:* Большинство вакансий находятся в г. Москва. Это может указывать на наличие большого рынка труда и возможностей для поиска работы в данном городе.

*Указание зарплаты в вакансиях:* Около половины размещенных вакансий имеют указанные границы зарплаты. Это может означать, что для значительной части вакансий зарплата является ключевой характеристикой, которая привлекает соискателей.

*Средние зарплаты:* Так как около половины размещенных вакансий имеют указанные границы зарплаты, средний уровень зарплаты 71 000 - 110 000 руб. может быть не совсем точен. Необходим дополнительный анализ для более точного определения средних значений.

*Тип занятости и зарплата:* Большинство вакансий размещено с полной занятостью. Это может указывать на стабильные позиции с предоставлением полного рабочего времени. Средняя заработная плата в диапазоне 71 000 - 110 000 руб. также может свидетельствовать о наличии вакансий с хорошими возможностями для заработка.

*Требования к опыту работы:* Большинство вакансий требуют опыт работы от 1 года до 3 лет. Также интересно, что вакансий без требования опыта работы больше, чем вакансий с требованиями опыта работы более 6 лет. Это может указывать на наличие возможностей для начинающих специалистов.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [281]:
# текст запроса
query_5_1 = f'''
        select e.name,
            count(v.id)
        from employers e
        join vacancies v on e.id = v.employer_id
        group by e.name
        order by 2 desc
        limit 5
        
'''

In [282]:
# результат запроса
cnt_vac_emp = pd.read_sql_query(query_5_1, connection)
display(cnt_vac_emp)

name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [283]:
# текст запроса
query_5_2 = f'''
        select a.name,
            count(e.id) cnt_employer,
            count(v.id) cnt_vacancies
        from areas a
        left join employers e on a.id = e.area
        left join vacancies v on a.id = v.area_id
        group by a.name
        having count(v.id) = 0
        order by count(e.id) desc
        limit 1
'''

In [284]:
# результат запроса
cnt_area = pd.read_sql_query(query_5_2, connection)
print()
print('Регион, в котором нет вакансий, но наибольшее количество работодателей: ', cnt_area.iloc[0,0])


Регион, в котором нет вакансий, но наибольшее количество работодателей:  Россия


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [285]:
# текст запроса
query_5_3 = f'''
        select e.name,
        count(distinct v.area_id) as regions_count
        from employers e
        left join vacancies v on e.id = v.employer_id
        group by e.name
        order by regions_count desc
'''

In [286]:
# результат запроса
cnt_area = pd.read_sql_query(query_5_3, connection)
cnt_area

name  regions_count
0                      Яндекс            181
1                  Ростелеком            152
2                  Спецремонт            116
3      Поляков Денис Иванович             88
4                   ООО ЕФИН              71
...                       ...            ...
23170                  СДЕЛКА              0
23171            Alandr Group              0
23172                     СДК              0
23173      Lemon Land Lombard              0
23174                  Махина              0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [287]:
# текст запроса
query_5_4 = f'''
        select count(e.id)
        from employers e
        left join employers_industries e_i on e.id = e_i.employer_id
        where e_i.industry_id is null
'''

In [288]:
# результат запроса
cnt_employer = pd.read_sql_query(query_5_4, connection)
print()
print('Количество работодателей, у которых не указана сфера деятельности', cnt_employer.iloc[0,0])


Количество работодателей, у которых не указана сфера деятельности 8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [289]:
# текст запроса
query_5_5 = f'''
        select e.name
        from employers e
        join employers_industries e_i on e.id = e_i.employer_id
        group by e.name
        having count(e_i.industry_id) = 4
        order by 1
        limit 1 offset 2
'''

In [290]:
# результат запроса
employer_industries = pd.read_sql_query(query_5_5, connection)
print()
print('Название компании, находящейся на третьем месте, у которых указано четыре сферы деятельности: ', employer_industries.iloc[0,0])


Название компании, находящейся на третьем месте, у которых указано четыре сферы деятельности:  2ГИС


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [291]:
# текст запроса
query_5_6 = f'''
        select count(e_i.employer_id)
        from industries i
        join employers_industries e_i on i.id = e_i.industry_id
        where i.name = 'Разработка программного обеспечения'
'''

In [292]:
# результат запроса
cnt_employ_po = pd.read_sql_query(query_5_6, connection)
print('Колчество работодателей со сферой деятльности - Разработка ПО: ', cnt_employ_po.iloc[0,0])

Колчество работодателей со сферой деятльности - Разработка ПО:  3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [293]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
table = (page.find('table', class_='standard sortable'))

cities = []
for row in table.find_all('tr'):
    cell = row.find_all('td')
    if len(cell)>=2:
        city = cell[1].text
        cities.append(city)

cities = tuple(cities)
# Выводим список городов-милионников
print()
print(cities)



('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [294]:
# текст запроса
query_5_7 = f'''
        with temp as (select a.name as region,
               count(v.id) count_yandex
        from areas a
        join vacancies v on v.area_id = a.id
        join employers e on e.id = v.employer_id
        where a.name in {cities} and e.name = 'Яндекс'
        group by a.name)
        
        select * from temp
        union all
        
        select 'Total', sum(count_yandex)
        from temp
        
'''

In [295]:
# результат запроса
yandex_vac = pd.read_sql_query(query_5_7, connection)
display(yandex_vac)

region  count_yandex
0         Волгоград          24.0
1           Воронеж          32.0
2      Екатеринбург          39.0
3            Казань          25.0
4         Краснодар          30.0
5        Красноярск          23.0
6            Москва          54.0
7   Нижний Новгород          36.0
8       Новосибирск          35.0
9              Омск          21.0
10            Пермь          25.0
11   Ростов-на-Дону          25.0
12           Самара          26.0
13  Санкт-Петербург          42.0
14              Уфа          26.0
15        Челябинск          22.0
16            Total         485.0

***

# Выводы по анализу работодателей
**Топ работодателей:**  Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть — самые активные компании по публикации вакансий. Это может указывать на их высокую популярность среди соискателей и/или на разнообразие предлагаемых вакансий.

**География вакансий:** Работодатели активно представлены в разных регионах. Некоторые компании даже публикуют вакансии в большом количестве регионов, что может свидетельствовать о их распределенности и стремлении привлечь таланты из разных уголков страны.

**Сферы деятельности:** Наличие большого числа компаний, не указавших сферу деятельности, может говорить о неполной информации в базе данных или о том, что часть компаний может быть менее структурированной.

**Города-милионники:** Компания Яндекс активно публикует вакансии в разных городах-милионниках, что подчеркивает ее распределенность и интерес к разнообразным кандидатам.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [296]:
# текст запроса
query_6_1 = f'''
        select count(id)
        from vacancies
        where name ilike '%data%' or name ilike '%данн%'
'''


In [297]:
# результат запроса
cnt_data_vac = pd.read_sql_query(query_6_1, connection)
print()
print('Количество вакансий, имеющих отношение к данным: ', cnt_data_vac.iloc[0,0])


Количество вакансий, имеющих отношение к данным:  1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [298]:
# текст запроса
query_6_2 = f'''
        select count(id)
        from vacancies
        where 
            (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or name ilike  '%machine learning%'
                or name ilike  '%машинн%обучен%'
                or (name not ilike '%HTML%' and name like '%ML%'))
            and (name ilike '%junior%'
                or experience ilike 'Нет опыта'
                or employment ilike 'Стажировка')
'''

In [299]:
# результат запроса
cnt_ds = pd.read_sql_query(query_6_2, connection)
print()
print('Количество вакансий для начинающего дата-сайентиста: ', cnt_ds.iloc[0,0])


Количество вакансий для начинающего дата-сайентиста:  51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [300]:
# текст запроса
query_6_3 = f'''
        select count(id)
        from vacancies
        where 
            (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or name ilike  '%machine learning%'
                or name ilike  '%машинн%обучен%'
                or (name not ilike '%HTML%' and name like '%ML%'))
            and (key_skills ilike '%SQL%' or key_skills ilike '%postgres%')

'''

In [301]:
# результат запроса
cnt_sql = pd.read_sql_query(query_6_3, connection)
print()
print('Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres: ',cnt_sql.iloc[0,0])


Количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres:  201


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [302]:
# текст запроса
query_6_4 = f'''
        select count(id)
        from vacancies
        where 
            (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or name ilike  '%machine learning%'
                or name ilike  '%машинн%обучен%'
                or (name not ilike '%HTML%' and name like '%ML%'))
            and (key_skills ilike '%Python%')

'''

In [303]:
# результат запроса
cnt_py = pd.read_sql_query(query_6_4, connection)
print()
print('Количество вакансий для DS, в которых в качестве ключевого навыка указан Python: ',cnt_py.iloc[0,0])


Количество вакансий для DS, в которых в качестве ключевого навыка указан Python:  351


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [304]:
# текст запроса
query_6_5 = f'''
        select round(avg(length(key_skills) - length(replace(key_skills, '\t', '')) + 1), 2)
        from vacancies
        where 
            (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or name ilike  '%machine learning%'
                or name ilike  '%машинн%обучен%'
                or (name not ilike '%HTML%' and name like '%ML%'))

'''

In [305]:
# результат запроса
cnt_skills = pd.read_sql_query(query_6_5, connection)
print()
print('Среднее количество ключевых навыков для вакансий DS: ',cnt_skills.iloc[0,0])


Среднее количество ключевых навыков для вакансий DS:  6.41


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [306]:
# текст запроса
query_6_6 = f'''
        select experience, round(avg(coalesce((salary_to + salary_from)/2, salary_from, salary_to, 0))) as avg_salary
        from vacancies
        where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or name ilike  '%machine learning%'
                or name ilike  '%машинн%обучен%'
                or (name not ilike '%HTML%' and name like '%ML%'))
                and (salary_to is not null or salary_from is not null)
        group by experience
        
'''

In [307]:
# результат запроса
avg_salary = pd.read_sql_query(query_6_6, connection)
print(avg_salary)

           experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0


***

# Выводы по предметному анализу
**Общее количество вакансий, связанных с наукой о данных, составляет 1771**. Это составляет около 3.6% от общего числа вакансий.

**Доступ для начинающих:** Наличие 51 вакансии для начинающих дата-сайентистов предоставляет шансы для новичков в этой сфере. Важно подчеркнуть, что начать карьеру в области данных возможно, хотя и с небольшим количеством доступных вакансий. 

**Широкий набор навыков:** Среднее количество ключевых навыков для вакансий (6.41) указывает на необходимость владения более чем одним компетентным навыком. Профессионалы в этой области должны быть готовы к использованию разнообразных инструментов.

**Зарплатные тренды:** Зарплаты для дата-сайентистов существенно меняются в зависимости от опыта работы. Более опытные специалисты могут ожидать более высоких заработных плат. Отсутсвие средней зп для опытных специалистов (с опытом более 6 лет) говорит о том, что в вакансиях не указана зп для этой категории людей

# Общий вывод по проекту

Данные представляют собой собранный объем информации о вакансиях, размещенных работодателями, которые содержат информацию о названии вакансии, работодателе, требованиях к соискателям, условиях занятости, городах и регионах размещения. Дополнительно, данные также содержат информацию о сфере деятельности, местоположении работодателя.

Анализ данных позволил выявить следующее:

Общее количество вакансий составляет 49 197, что указывает на большую выборку для анализа.
Наличие разнообразных сфер деятельности и регионов говорит о разносторонности и географическом охвате предоставленных вакансий.
Большинство вакансий размещаются в Москве, что акцентирует внимание на крупных рынках труда.
Наличие указанных зарплатных ожиданий и требований к опыту работы делает данные полезными для анализа и рекомендаций.
Высокий спрос на аналитические навыки, в частности связанные с наукой о данных, подчеркивает актуальность целевой позиции "Data Scientist".
Таким образом, данные достаточно разнообразны, содержат важные параметры для анализа вакансий в области Data Scientist и могут быть использованы для создания модели машинного обучения, направленной на рекомендацию подходящих вакансий клиентам агентства.

***Дополнительные исследования, которые могут быть проведены:***

**Анализ сферы деятельности работодателей:** В каких сферах деятельности наиболее активно публикуются вакансии для дата-сайентистов. Это поможет понять, какие отрасли в настоящее время наиболее нуждаются в аналитических специалистах. А также выведем ТОП-5 компаний по количеству вакансий для дата-сайентистов.

In [331]:
query_6_7 = f'''
      select i.name, count(*) as vacancy_count
from vacancies v
left join employers_industries e_i on e_i.employer_id = v.employer_id
left join industries i on e_i.industry_id = i.id
where 
    (v.name ilike '%data scientist%'
    or v.name ilike '%data science%'
    or v.name ilike '%исследователь данных%'
    or v.name ilike '%machine learning%'
    or v.name ilike '%машинн%обучен%'
    or (v.name not ilike '%HTML%' AND v.name like '%ML%'))
group by i.name
order by vacancy_count desc
limit 15
'''

In [333]:
# результат запроса
dop = pd.read_sql_query(query_6_7, connection)
print()
print(dop)


                                                                                                                                                name   
0                                                                                                                Разработка программного обеспечения  \
1                                                 Системная интеграция,  автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг   
2                                                                                                                                               Банк   
3   Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее)   
4                                                                                                                                               None   
5                                                                                      

**Разнообразие отраслей:** Вакансии для дата-сайентистов и аналитиков данных публикуются в разнообразных отраслях. Важно заметить, что специалисты по анализу данных востребованы не только в технологических компаниях, но также в сферах, связанных с банковскими услугами, интернет-компаниями, маркетингом, управлением активами, консалтингом и других.

Наивысшее количество вакансий имеют "Разработка программного обеспечения" и "Системная интеграция, автоматизации технологических и бизнес-процессов предприятия, ИТ-консалтинг". Это указывает на высокий спрос на специалистов в этих отраслях, где аналитические навыки играют важную роль для разработки программных продуктов, интеграции систем и консультирования в области информационных технологий.

In [340]:
query_6_8 = f'''
select e.name, count(*) as vacancy_count
from vacancies v
left join employers e on e.id = v.employer_id
where 
    (v.name ilike '%data scientist%'
    or v.name ilike '%data science%'
    or v.name ilike '%исследователь данных%'
    or v.name ilike '%machine learning%'
    or v.name ilike '%машинн%обучен%'
    or (v.name not ilike '%HTML%' AND v.name like '%ML%'))
group by e.name
order by vacancy_count desc
limit 5
'''

In [341]:
# результат запроса
dop_1 = pd.read_sql_query(query_6_8, connection)
print()
print(dop_1)


                    name  vacancy_count
0                   СБЕР             37
1        Bell Integrator             25
2         Банк ВТБ (ПАО)             18
3                     VK             15
4  Positive Technologies             11


ТОП - 5 компаний по спросу дата-сайентистов. Компания "СБЕР" имеет самое высокое количество вакансий для дата-аналитиков среди представленных в списке.